### ISSUE TITLE
Spaces in LP file format variable names are not skipped

### ISSUE THREAD

**Issue by fuglede** [ERGO-Code/HiGHS#851]
According to the [file format specification](http://lim.univ-reunion.fr/staff/fred/Enseignement/Optim/doc/CPLEX-LP/CPLEX-LP-file-format.html), "In general, white space between characters is irrelevant as it is skipped when a file is read", and the specification goes on with the example that the constraint `x1 x2 = 0` is parsed as a variable called "x1x2" being 0.

That's not what HiGHS does, though: As far as I can tell, it silently converts `x1 x2` into `x1 + x2`. Here's an example:

```
$ cat space.lp
Maximize
  x1 x2
Subject to
  x1 x2 = 0
End

$ highs space.lp --solution_file sol
Running HiGHS 1.2.2 [date: 2022-05-04, git hash: d1d9839b-dirty]
Copyright (c) 2022 ERGO-Code under MIT licence terms
LP   space has 1 rows; 2 cols; 2 nonzeros
Presolving model
0 rows, 0 cols, 0 nonzeros
0 rows, 0 cols, 0 nonzeros
Presolve : Reductions: rows 0(-1); columns 0(-2); elements 0(-2) - Reduced to empty
Solving the original LP from the solution after postsolve
Model   status      : Optimal
Objective value     :  0.0000000000e+00
HiGHS run time      :          0.00

$ cat sol
Model status
Optimal

# Primal solution values
Feasible
Objective 0
# Columns 2
x1 0
x2 0
# Rows 1
R0 0
```

I ran into this problem after trying to read an LP file exported by Xpress, whose file writer can include spaces in variable names for readability.

A more realistic example than the above is the following, which causes HiGHS to throw a parser error:

```
Maximize
 obj: Q( a )
Bounds
 0 <= Q( a ) <= 40
End
```

A variation of the above that for some reason works is

```
Maximize
 obj: Q(a )
Bounds
 0 <= Q(a ) <= 40
End
```

which creates two variables named `Q(a` and `)`.

This might be related to #846.

**Comment by fuglede** [comment]
Spaces in constraint names are also problematic:

```
$ cat test.lp
Maximize
  x
Subject to
  some ctr: x <= 40
End

$ highs test.lp
Running HiGHS 1.2.2 [date: 2022-05-04, git hash: d1d9839b-dirty]
Copyright (c) 2022 ERGO-Code under MIT licence terms
ERROR:   Parser error reading test.lp
Error loading file
```

**Comment by jajhall** [comment]
I can well imagine that some edge cases in handling .lp format aren't covered. TBH, anyone who puts spaces in variable or constraint names is asking for trouble!

**Comment by fuglede** [comment]
> TBH, anyone who puts spaces in variable or constraint names is asking for trouble!

FWIW, this came about from using the C# bindings for Xpress, whose `XPRSProb.exportProb` happily spits out anything you throw into variable names, including invalid characters such as `-`. Spaces are sort of useful for readability ("`Price(object A) <= 60`") but edge-casey indeed.

The [tokenizer](https://github.com/ERGO-Code/HiGHS/blob/master/extern/filereaderlp/reader.cpp#L983) already skips whitespace, so it looks like the groundwork has been done. If you like, I can also look into why that doesn't quite suffice.

**Comment by jajhall** [comment]
Once names have been defined by a modeller, it's not for a solver interface to modify them. I guess that these days people are increasingly remote from the practicalities of model definition and parsing computer-generated results with code, so defining variables as phrases containing spaces doesn't sound so odd! It's still not a priority for HiGHS to handle this at the moment, unless the fix is straightforward

**Comment by fuglede** [comment]
So, `readnexttoken` effectively splits on a number of predefined characters, including space. Excluding space from the list and instead removing spaces by hand is _almost_ enough, except spaces do have a meaning for keywords, so that case needs to be handled explicitly. This can be done by using `parsesectionkeyword` but it looks like that one is intended to be used only once the tokens are in place, so that's a little awkward.

But from a quick look, this means that something like the below would suffice:

```diff
diff --git a/extern/filereaderlp/reader.cpp b/extern/filereaderlp/reader.cpp
index 84eebbb4..24b0965a 100644
--- a/extern/filereaderlp/reader.cpp
+++ b/extern/filereaderlp/reader.cpp
@@ -2,6 +2,7 @@

 #include "builder.hpp"

+#include <algorithm>
 #include <cassert>
 #include <iostream>
 #include <fstream>
@@ -1008,11 +1011,17 @@ void Reader::readnexttoken() {
    }

    // assume it's an (section/variable/constraint) identifier
-   auto endpos = this->linebuffer.find_first_of("\t\n\\:+<>^= /-*", this->linebufferpos);
-   if( endpos == std::string::npos )
+   auto endpos = this->linebuffer.find_first_of("\t\n\\:+<>^=/-*", this->linebufferpos);
+   if ( endpos == std::string::npos )
       endpos = this->linebuffer.size();  // take complete rest of string
-   if( endpos > this->linebufferpos ) {
-      this->rawtokens.push_back(std::unique_ptr<RawToken>(new RawStringToken(std::string(this->linebuffer, this->linebufferpos, endpos - this->linebufferpos))));
+   if ( endpos > this->linebufferpos ) {
+      std::string rest = std::string(this->linebuffer, this->linebufferpos, endpos - this->linebufferpos);
+      // Remove spaces from the rest of the string as long as this is not a keyword
+      if (parsesectionkeyword(rest) == LpSectionKeyword::NONE) {
+         rest.erase(std::remove(rest.begin(), rest.end(), ' '), rest.end());
+      }
+      this->rawtokens.push_back(std::unique_ptr<RawToken>(new RawStringToken(rest)));
       this->linebufferpos = endpos;
       return;
    }
```

**Comment by feldmeier** [comment]
supporting spaces in constraint names should be relatively easy to add, since that's the only occasion where the colon character is being used.

spaces in variables names are nontrivial, and it might require quite a bit of rework on the filereader to get right. I don't have time to work on this right now; I'd be happy to review a pull request implementing the changes necessary. Otherwise I'll add it to my backlog and work on it once I find the time.

**Comment by jajhall** [comment]
I'm happy to state that we don't support spaces in variable names for now, even if .lp is more important than it seemed before. 

Spotting them and bailing out cleanly would be nice. The guy from Siemens Energy had HiGHS hang before I realised that he'd not got a newline after the last constraint. It's fixed in master, but he was using binaries.

**Comment by odow** [comment]
The real problem is that people are using LP files!!! Whhyyyyyyy.

**Comment by jajhall** [comment]
Because they don't use modelling languages and it's the easiest way to export a model if you've written it in code. 

This week at least I converted an industrial client from calling lp_solve with their LP file into calling HiGHS with a Cplex LP file! 😂

**Comment by odow** [comment]
> Because they don't use modelling languages

1983 is calling: https://dl.acm.org/doi/pdf/10.1145/357456.357457

:smile: but I'm mostly sarcasm. I do see the utility in passing LP files.
